extensions/amp-animation/0.1/web-animations.js

```diff
@@ -0,0 +1,363 @@
+ */
+class Scanner {
+
+  /**
+   * @param {!WebAnimationDef|!Array<!WebAnimationDef>} spec
+   */
+  scan(spec) {
+    if (isArray(spec)) {
+      spec.forEach(spec => this.scan(spec));
+      return;
+    }
+
+    // WebAnimationDef: (!WebMultiAnimationDef|!WebKeyframeAnimationDef)
+    if (spec.animations) {
+      this.onMultiAnimation(/** @type {!WebMultiAnimationDef} */ (spec));
+    } else if (spec.keyframes) {
+      this.onKeyframeAnimation(/** @type {!WebKeyframeAnimationDef} */ (spec));
+    } else {
+      this.onUnknownAnimation(spec);
+    }
+  }
+
+  /** @param {!WebMultiAnimationDef} unusedSpec */
+  onMultiAnimation(unusedSpec) {
+    dev().assert(null, 'not implemented');
+  }
+
+  /** @param {!WebKeyframeAnimationDef} unusedSpec */
+  onKeyframeAnimation(unusedSpec) {
+    dev().assert(null, 'not implemented');
+  }
+
+  /** @param {!Object} unusedSpec */
+  onUnknownAnimation(unusedSpec) {
+    throw dev().createError('unknown animation type:' +
+        ' must have "animations" or "keyframes" field');
+  }
+}
+
+
+/**
+ * The scanner that evaluates all expressions and builds the final
+ * `WebAnimationRunner` instance for the target animation. It must be
+ * executed in the "measure" vsync phase.
+ */
+export class MeasureScanner extends Scanner {
+
+  /**
+   * @param {!Window} win
+   * @param {{
+   *   resolveTarget: function(string):?Element,
+   * }} context
+   * @param {boolean} validate
+   */
+  constructor(win, context, validate) {
+    super();
+    /** @const */
+    this.win = win;
+
+    /** @const @private */
+    this.context_ = context;
+
+    /** @const @private */
+    this.validate_ = validate;
+
+    /** @private {!Array<!InternalWebAnimationRequestDef>} */
+    this.requests_ = [];
+
+    /** @private {!WebAnimationTimingDef} */
+    this.timing_ = {
+      duration: 0,
+      delay: 0,
+      endDelay: 0,
+      iterations: 1,
+      iterationStart: 0,
+      easing: 'linear',
+      direction: WebAnimationTimingDirection.NORMAL,
+      fill: WebAnimationTimingFill.NONE,
+    };
+
+    /** @private {!Array<!Element> } */
+    this.targets_ = [];
+
+    /** @private {!Array<!CSSStyleDeclaration>} */
+    this.computedStyleCache_ = [];
+  }
+
+  /** @return {!WebAnimationRunner} */
+  createRunner() {
+    return new WebAnimationRunner(this.requests_);
+  }
+
+  /** @override */
+  onMultiAnimation(spec) {
+    const newTiming = this.mergeTiming_(spec, this.timing_);
+    this.withTiming_(newTiming, () => {
+      this.scan(spec.animations);
+    });
+  }
+
+  /** @override */
+  onKeyframeAnimation(spec) {
+    const timing = this.mergeTiming_(spec, this.timing_);
+    const target = this.resolveTarget_(spec.target);
+    /** @type {!WebKeyframesDef} */
+    let keyframes;
+
+    if (isObject(spec.keyframes)) {
+      // Property -> keyframes form.
+      const object = /** {!Object<string, *>} */ (spec.keyframes);
+      keyframes = {};
+      for (const prop in object) {
```

Comment:
This, and below, will allow any animatable property to be animated. Is validation that only opacity and transform animate happen somewhere else?